ctlearn/image_mapping.py

```diff
@@ -143,25 +150,83 @@ def map_image(self, pixels, telescope_type):
+                    # Get relevant parameters
+                    output_dim = self.image_shapes[telescope_type][0]+pad*2
+                    num_pixels = self.num_pixels[telescope_type]
+                    pixel_length = self.pixel_lengths[telescope_type]
+                    '''
+                    The pixel positions from files are already rotated!
+                    if telescope_type in ["LST", "MSTN"]:
+                        pos = self.rotate_cam(pos)
+                    '''
+                    # Adding intensities (0.0) for virtual pixels outside the camera
+                    x=pos[0]
+                    y=pos[1]
+                    z=vector[1:num_pixels+1]
+                    while x.shape!=z.shape:
+                        z=np.concatenate((z,[0.0]))
+                    # Shifting and rescaling the pixel positions
+                    x -= np.min(x)
+                    y -= np.min(y)
+                    pad*=2
+                    x *= ((output_dim-pad)/np.max(x))
+                    y *= ((output_dim-pad)/np.max(y))
+                    if (telescope_type in ['MGC']):
+                        x += pad
+                        y += pad
+                    else:
+                        x += pad/2
+                        y += pad/2
+                    total_intensity = np.sum(z)
+                    # Creating the output grid.
+                    # Change thrid argument in np.linespace for different resolution.
+                    ti = np.linspace(0, output_dim, output_dim)
+                    XI, YI = np.meshgrid(ti, ti)
+                    # Applying the interpolation
+                    if self.hex_conversion_algorithm == 'nearest_interpolation':
+                        image_2D = griddata((x.ravel(), y.ravel()), z.ravel(), (XI, YI), method='nearest',fill_value=0.0)
+                        interpolated_total_intensity = np.sum(image_2D)
+                        image_2D *= (total_intensity/interpolated_total_intensity)
+                    elif self.hex_conversion_algorithm == 'linear_interpolation':
+                        image_2D = griddata((x.ravel(), y.ravel()), z.ravel(), (XI, YI), method='linear',fill_value=0.0)
+                        interpolated_total_intensity = np.sum(image_2D)
+                        image_2D *= (total_intensity/interpolated_total_intensity)
+                    elif self.hex_conversion_algorithm == 'cubic_interpolation':
+                        image_2D = griddata((x.ravel(), y.ravel()), z.ravel(), (XI, YI), method='cubic',fill_value=0.0)
+                        interpolated_total_intensity = np.sum(image_2D)
+                        image_2D *= (total_intensity/interpolated_total_intensity)
+                    image_2D = np.expand_dims(image_2D, axis=2)
+                else:
+                    raise ValueError("Sorry! Telescope type {} isn\'t supported with the conversion algorithm \'{}\'.".format(telescope_type,self.hex_conversion_algorithm))
```

Comment:
This check differs from the one on line 88, because it's checking if the hex_conversion_algorithm is available for a specific telescope type. For the cameras with rectangle pixels (MSTS, SSTC, SSTA) the hex_conversion_algorithm isn't available. With the previous checks the code won't break, if you select MSTS with linear_interpolation.

Reply:
Gotcha, that makes sense.

Reply:
@nietootein @aribrill If we want to run the whole CTA array (LST's, MST's and SST's together) using the new interpolation methods, the code will break here. Would it be better to convert the conversion algorithm to oversampling for the rectangle pixels cameras (MSTS, SSTC, SSTA) and not break?

Reply:
Good point. Maybe it makes sense to be explicit and have two separate options for hexagonal_mapping_method and square_mapping_method.

Reply:
Is it possible that we might want to be even more specific and allow ourselves to set a conversion method (whether square or hexagonal) for each telescope type separately (like we already do for padding)? It seems plausible to me (although maybe you might disagree) that you could want to use a higher-quality (but more time-consuming) hex conversion algorithm for some telescopes and a cheaper, lower-quality one for others, all within the same run. For example, you might discover that on the MSTs, higher-fidelity conversion somehow improves results, but on LSTs it doesn't.

Reply:
(Sorry for late reply.) I think it will be possible (minor changes in the code). Does the image shape size of each camera type must be the same, if we train in stereo type? (This is what i thought of.) If this is the case, we need to have the same quality/resolution for each camera type. Therefore, we need to add new interpolation methods for the rectangle camera type. Let's discuss in the call tomorrow. Maybe opening an issue on this would also be good.

Reply:
I don't think it's necessary to enforce that each method result in the same final image shape, as long as it's possible to manually set this to be the case with the right choice of settings (we could make those settings the defaults for convenience). Depending on the user's network design, it may or may not be required that the image shape be invariant to the mapping method. We can discuss tomorrow.